### PR TITLE
Background model

### PR DESCRIPTION
**Pull request by guynir42**
Adds a Background object that saves either a map or a scalar (and later we can add a polynomial fit model) to disk and to database, that holds the information about what background needs to be subtracted from the Image. 

This happens along with source detection and PSF finding, as part of the "extraction" step. 

The background object joins the other image products: sources, psf, wcs, zp and now bg. 

When a user wants to get an image that has its background subtracted, there would be a method in Image that uses this object to subtract the background estimate from the image (either a map of the same size as the image data, or a scalar). 

If the image is already background free (e.g., it is a subtraction) then we save a "zero" background which has a single scalar value of zero, and its subtraction is a no-op. 

Note that in this case a background object can still maintain an estimate of the background variance, if we can get it.

### CODE REVIEW

**Comment by guynir42** [comment]
This PR addresses mainly the background model, as in issue #205. 
While I'm at it, I'm also fixing some issues that have to do with extraction. 
- The sextractor thresholds based on the PSF shape (issue #220 )
- The apertures using the config file, rather than the instrument (issue #304 )
- Ability to turn off using a local background estimate when running photometry on cutouts (issue #230 )
- Subtract background when coadding images (issue #186 )
- Configurable sky subtraction (issue #51 )
- Make sure background estimates get into the Image object upon extracting sources (issue #173 )
- I've removed that last test in `test_extraction` as it was very similar to the one before it, and seemed redundant (issue #164 )

File: models/measurements.py [hunk 1]
```diff
@@ -165,15 +166,15 @@ def mag_apertures_err(self):
 
     @property
     def magnitude(self):
+        mag = -2.5 * np.log10(self.flux) + self.zp.zp
         if self.best_aperture == -1:
-            return self.mag_psf
-        return self.mag_apertures[self.best_aperture]
+            return mag
+        else:
+            return mag + self.zp.aper_cors[self.best_aperture]
```
Comment (rknop) [inline models/measurements.py:173]: ...this probably obviates my previous comment...

**Review by rknop — APPROVED** [review]
This looks good.  Only change I want is in the docstring on Background.save() in models/background.py (HDF5 dataset names in the docstring aren't up to date with what's actually written, plus I want to make it anally clear exactly what they are).

File: models/background.py [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+        """Write the Background to disk.
+
+        May or may not upload to the archive and update the
+        FileOnDiskMixin-included fields of this object based on the
+        additional arguments that are forwarded to FileOnDiskMixin.save.
+
+        This saves an HDF5 file that contains a single group called "/background".
+        It will have a few attributes, notably: "format", "value", "noise" and "image_shape".
+
+        If the format is "map", there are two datasets under this group: "background/mean" and "background/var".
```
[the remark this answers is not in the document]
Comment (guynir42) [reply models/background.py:216]: I'll refactor the docstrings.

File: models/background.py [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+                    raise RuntimeError(
+                        f"Counts shape {self.counts.shape} does not match image shape {self.image_shape}"
+                    )
+                if self.variance.shape != self.image_shape:
+                    raise RuntimeError(
+                        f"Variance shape {self.variance.shape} does not match image shape {self.image_shape}"
+                    )
+
+                bggrp.create_dataset( 'counts', data=self.counts )
+                bggrp.create_dataset( 'variance', data=self.variance )
```
[the remark this answers is not in the document]
Comment (guynir42) [reply models/background.py:288]: I think the right way to put it is "it is an attempt to get an estimate of the sky (mean/rms) alone, not including sources, but often the sources do bias the sky estimation if they are very bright".

File: models/image.py [hunk 4]
```diff
@@ -1999,6 +2015,26 @@ def nanscore(self):
     def nanscore(self, value):
         self._nanscore = value
 
+    @property
+    def data_bg(self):
```
[the remark this answers is not in the document]
Comment (guynir42) [reply models/image.py:2019]: Yes, data_bgsub is better.

File: models/measurements.py [hunk 5]
```diff
@@ -117,18 +117,19 @@ class Measurements(Base, AutoIDMixin, SpatiallyIndexed, HasBitFlagBadness):
     def flux(self):
         """The background subtracted aperture flux in the "best" aperture. """
         if self.best_aperture == -1:
-            return self.flux_psf - self.background * self.area_psf
+            return self.flux_psf - self.bkg_mean * self.area_psf
         else:
-            return self.flux_apertures[self.best_aperture] - self.background * self.area_apertures[self.best_aperture]
+            return self.flux_apertures[self.best_aperture] - self.bkg_mean * self.area_apertures[self.best_aperture]
```
[the remark this answers is not in the document]
Comment (guynir42) [reply models/measurements.py:122]: I think it is understood that flux is not corrected in any way. The zero point is a translation of flux to magnitude so if you correct anything, it is only on the magnitudes.

**Pull request merged**